src/LSST/faultmgr/BasicFaultMgr.h

```diff
@@ -0,0 +1,218 @@
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#ifndef LSST_M2CELLCPP_BASICFAULTMGR_H
+#define LSST_M2CELLCPP_BASICFAULTMGR_H
+
+// System headers
+#include <functional>
+#include <map>
+#include <memory>
+#include <mutex>
+
+// Project headers
+#include "control/control_defs.h"
+#include "faultmgr/FaultStatusBits.h"
+#include "util/clock_defs.h"
+
+namespace LSST {
+namespace m2cellcpp {
+namespace faultmgr {
+
+class FaultMgr;
+
+/// Stores a copy of fault information for a specific system.
+/// All systems use the same layout for a uint64_t with the bits mapped
+/// to specific faults (see `FaultStatusBits`).
+/// The purpose of most of the member variables is not well defined.
+/// `_currentFaults` is a record of the most recently set faults and it
+///    can be completely overwritten during a fault update.
+/// `_summaryFaults` is a record of all faults that haven't been reset.
+///    Values are usually latched to '1' and usually only go back to '0'
+///    when `resetFaults()` is called.
+/// `_prevFaults` is usually the previous value of `_currentFaults`,
+///    which seems to be used to limit message passing in LabView. This
+///    isn't that important in the C++ code and should probably be removed.
+/// `_faultEnableMask` is used enable/disable faults.
+/// `_defaultFaultMask` may be used for the default enable mask.
+/// `_affectedFaultsMask` and `_affectedWarnInfoMask` in subsystems
+/// appear to indicate which fault bits cause faults or warnings for
+/// the subsystem. The central copy has all of the local `_affected`
+/// masks added to its `_affected` masks during updates (see
+/// `fault_manager_main.vi` "User Event:FaultsTelem").
+///
+/// They shouldn't need to have separate copies, as faults are system
+/// wide, but the LabView code has a central copy and then one copy per
+/// subsystem (most of them anyway) that are kept in sync using
+/// "FaultsTelem" User Event messages. There are subtle inconsistencies
+/// between how they are implemented in the LabView code and it's
+/// difficult to tell what effect the inconsistencies have, or
+/// are meant to have.
+/// The message passing has been eliminated but extra copies
+/// still remain.
+/// "BasicFaultManager.lvclass:init.vi"
+class BasicFaultMgr {
+public:
+
+    /// cRIO subsystem enum values from SystemControlDefs.lvlib:cRIO_Subsystem.ctl
+    /// Note: This may not be the best place to put this.
+    enum CrioSubsystem {
+        SYSTEM_CONTROLLER = 0, ///< "System controller" = 0
+        FAULT_MANAGER = 1,     ///< "Fault Manager" = 1
+        POWER_SUBSYSTEM = 2,   ///< "Power Subsystem" = 2
+        CELL_CONTROLLER = 3,   ///< "Cell Controller" = 3
+        TELEMETRY_LOGGER = 4,  ///< "Telemetry Logger" = 4
+        NETWORK_INTERFACE = 5, ///< "Network Interface" = 5
+        MOTION_ENGINE = 6      ///< "Motion Engine" = 6
+    };
```

Comment:
In python, usually, we might have a `enums.py` to organize the `enums` in the project. But in C++, do the people do the similar things? Or C++ developers are more used to go to each header/class files to look for the possible enums? I just felt this is inefficient if you have >10 classes.

Reply:
C++ it is normal to have the enums in headers and it is where they are expected to be found. I'm not happy with the current location of these and will move them, but I'm not sure where I want to put them yet.

It would probably be a good idea to verify these against what the GUI expects. This can be done with a file, but there are issues with the file not being the right version, multiple versions, . Another way would be to make a json message, so that the controller can send the bitmap values to the GUI (or whatever is connected to tcp/ip) and verify that their bitmapped values match.